### DISCUSSION TITLE
The most "appropriate" way to match `libp2p::TransportError::Other`?

### DISCUSSION THREAD

**Discussion by gnattu** [libp2p/rust-libp2p#5114] [Q&A] [answered]
I'm trying to make error handling logic for `SwarmEvent::OutgoingConnectionError`. But the `libp2p::TransportError::Other` is extremely hard to match because it comes from all over the place and I don't know how to match them exactly.

For example, the timeout Error is printed like `Other(Custom { kind: Other, error: Timeout })`, and connection refused is printed like `Other(Custom { kind: Other, error: Other(Left(Left(Os { code: 61, kind: ConnectionRefused, message: "Connection refused"  }))) })` 

I currently use `format!` to turn them into string and check if the string contains anything like `Timeout` or `ConnectionRefused`, but is there any practical way that I can do this better? I know that we can use `get_ref()` but I don't know which type should I down_cast into after that.

**Comment by thomaseizinger** [comment]
Walking the source chain is likely the best option. Most errors will be `io::Error`s. If you are looking for something specific, then it depends on what transports you are using, e.g. if you are using QUIC, then the error may also be a `libp2p::quic::Error`.

But before you go down that rabbit hole, can you explain why you need to specific error that occurred? Maybe your problem can be solved in a different way.

**Reply by gnattu** [reply]
I want to check if the outgoing connection error is due to our(the dialer) side or is it really the remote peer is not reachable. If it's the former I will schedule a re-dial a few seconds later, if it's the latter I will just remove that peer from the routing table. Only knowing it is an IO Error is not enough because we may have a case where we already have an address up but that address is not able to connect to the Internet yet, if that is the case, I will get an `NetworkUnreachable` Error so I know that the networking is not ready on our side. But if it is `ConnectionRefused` or `Timeout`, then it is probably the remote side.

**Reply by thomaseizinger** [reply, marked as answer]
In that case, I think your best bet is to walk the source chain of the error, try to downcast each one to an IO error and checking its `ErrorKind`.

In general, this seems like a very fragile way of going about it though. Also, note that we will automatically remove every address that generated a dial error from the routing table already:

https://github.com/libp2p/rust-libp2p/blob/1139f1c170c36d7223bb8c52c667b111ea5c5fbc/protocols/kad/src/behaviour.rs#L2001-L2005
